lib/elixir/src/elixir_erl_var.erl

```diff
@@ -114,7 +114,8 @@ should_warn(Meta) ->
 mergev(S1, S2) ->
   S2#elixir_erl{
     vars=merge_vars(S1#elixir_erl.vars, S2#elixir_erl.vars),
-    export_vars=merge_opt_vars(S1#elixir_erl.export_vars, S2#elixir_erl.export_vars)
+    export_vars=merge_opt_vars(S1#elixir_erl.export_vars, S2#elixir_erl.export_vars),
+    ssa_types=maps:merge(S1#elixir_erl.ssa_types, S2#elixir_erl.ssa_types)
```

Comment:
While this is enough to merge types across clauses, you still need to clean up clauses for any conflicting type as mentioned in another comment above.